update/v1alpha1/types_update_status.go

```diff
@@ -0,0 +1,750 @@
+	// +optional
+	Conditions []metav1.Condition `json:"conditions,omitempty"`
+}
+
+// UpdateInformer is an insight producer identified by a name, carrying a list of insights it produced
+type UpdateInformer struct {
+	// Name is the name of the insight producer
+	// +required
+	// +kubebuilder:validation:Required
+	Name string `json:"name"`
+
+	// Insights is a list of insights produced by this producer
+	// +optional
+	Insights []UpdateInsight `json:"insights,omitempty"`
+}
+
+// ControlPlaneAssessment is the assessment of the control plane update process
+type ControlPlaneAssessment string
+
+const (
+	// Unknown means the update status and health cannot be determined
+	ControlPlaneAssessmentUnknown ControlPlaneAssessment = "Unknown"
+	// Progressing means the control plane is updating and no problems or slowness are detected
+	ControlPlaneAssessmentProgressing ControlPlaneAssessment = "Progressing"
+	// Completed means the control plane successfully completed updating and no problems are detected
+	ControlPlaneAssessmentCompleted ControlPlaneAssessment = "Completed"
+	// Degraded means the process of updating the control plane suffers from an observed problem
+	ControlPlaneAssessmentDegraded ControlPlaneAssessment = "Degraded"
+)
+
+// ClusterVersionStatusInsightConditionType are types of conditions that can be reported on ClusterVersion status insight
+type ClusterVersionStatusInsightConditionType string
+
+const (
+	// Updating condition communicates whether the ClusterVersion is updating
+	ClusterVersionStatusInsightUpdating ClusterVersionStatusInsightConditionType = "Updating"
+)
+
+// ClusterVersionStatusInsightUpdatingReason are well-known reasons for the Updating condition on ClusterVersion status insights
+type ClusterVersionStatusInsightUpdatingReason string
+
+const (
+	// CannotDetermineUpdating is used with Updating=Unknown
+	ClusterVersionCannotDetermineUpdating ClusterVersionStatusInsightUpdatingReason = "CannotDetermineUpdating"
+	// ClusterVersionProgressing means that ClusterVersion is considered to be Updating=True because it has a Progressing=True condition
+	ClusterVersionProgressing ClusterVersionStatusInsightUpdatingReason = "ClusterVersionProgressing"
+	// ClusterVersionNotProgressing means that ClusterVersion is considered to be Updating=False because it has a Progressing=False condition
+	ClusterVersionNotProgressing ClusterVersionStatusInsightUpdatingReason = "ClusterVersionNotProgressing"
+)
+
+// VersionMetadataType is the type of the metadata value
+// +kubebuilder:validation:Enum=string;bool;int
+type VersionMetadataType string
+
+const (
+	StringVersionMetadata VersionMetadataType = "string"
+	BoolVersionMetadata   VersionMetadataType = "bool"
+	IntVersionMetadata    VersionMetadataType = "int"
+)
+
+type VersionMetadataKey string
+
+const (
+	// installation denotes a boolean that indicates the update was initiated as an installation
+	InstallationMetadata VersionMetadataKey = "installation"
```

Comment:
`ControlPlaneUpdateVersions` has separate `previous` and `target`.  Do we need an explicit "I'm the install!" marker in version metadata?  Or can that be inferred from "`previous` is empty"?  I can't think of a situation where you'd be in the middle of a real A->B update but still care about whether A was an install version or not.

Reply:
this basically mimics what we do in `oc` prototype:

https://github.com/openshift/oc/blob/3692450b96d57ae3870e5f693e833a0701d0e2b0/pkg/cli/admin/upgrade/status/controlplane.go#L57-L61

if this was the only use case for version metadata then maybe the implicit 'previous empty implies installation' would be good enough but we have a good use case for metadata (partial) so I think it's worth being explicit, we have the mechanism for it

Reply:
I have no problem with openshift/oc#4978e7d226e8cb2 deciding to use an explicit property internally to read more clearly.  And I'm fine with you following that pattern here in the initial v1alpha1 approach.  But the public-API form increases my personal threshold for whether the structure should allow inconsistent data, and "claims to be an `installation` but has a different `previous`" and "claims to be an update, but has no `previous`" are two possible inconsistent states for this structure.  But either way, feel free to mark this thread resolved, to get v1alpha1 landed :)

Reply:
I like that point. Not having more metadata is also simpler which is always a better start. I'll change this.